docs/design/2023-11-29-priority-queue-for-auto-analyze.md

```diff
@@ -0,0 +1,270 @@
+[autoanalyze.go?L287:6]: https://sourcegraph.com/github.com/hi-rustin/tidb@d4618d4a5f91ee3703336fd1ba328c2e477652e5/-/blob/pkg/statistics/handle/autoanalyze/autoanalyze.go?L287:6&popover=pinned
+[autoanalyze.go?L301:40]: https://sourcegraph.com/github.com/hi-rustin/tidb@d4618d4a5f91ee3703336fd1ba328c2e477652e5/-/blob/pkg/statistics/handle/autoanalyze/autoanalyze.go?L301:40&popover=pinned
+[autoanalyze.go?L262:15]: https://sourcegraph.com/github.com/hi-rustin/tidb@d4618d4a5f91ee3703336fd1ba328c2e477652e5/-/blob/pkg/statistics/handle/autoanalyze/autoanalyze.go?L262:15&popover=pinned
+
+The above process is the current auto analyze process. We can see that the current auto analyze process is a random selection algorithm. We need to design a priority queue to solve the above problems.
+
+Since we perform analysis tasks synchronously on a single point for each table, we need to carry out weighted sorting on the tables that need analysis.
+
+   1. We still use tidb_auto_analyze_ratio to determine if the table needs to be analyzed. The default value is 0.5. It means that auto-analyze is triggered when greater than 50% of the rows in a table have been modified.
+   2. For special events, it is necessary to put the table into the priority queue. For example, the table has a new index but it hasn't been analyzed yet.
+   3. If a table is added to the auto-analysis queue, a weight must be assigned to it to determine the execution order.
+
+Weight table:
+| Name                 | Meaning                                                                 | Weight                                                                                                                                                                                                                                                                                                                                                            |
+|----------------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Percentage of Change | The percentage of change since the last analysis.                       | 50% <= x < 70%: 1 <br/> 70% <= x < 80%: 1.3 <br/> x >= 80%: 1.5  <br/>                                                                                                                                                                                                                                                                                            |
```

Comment:
Besides, some other things can also be taken into consideration.

1. Is it partition table?

partition table will analyze every partition and do ```merge```. so it will take more time.

3. NDV

after this [PR](https://github.com/pingcap/tidb/pull/48902), the low NDV table will analyze faster than the high NDV table.

Reply:
> 1. Is it partition table?

During the meeting, we discussed this indicator that is related to the size of a table. However, we don't want to focus solely on the size of the table because even small tables can lead to some really bad plans for certain join queries. Instead, we want to focus on the `Percentage of Change` of the table. Even if we give a partition table a higher priority, it doesn't solve the issue of resource starvation. To address this problem, we need to support the execution of analysis tasks on different nodes. If we continue to execute tasks on a single node, even with complicated priority rules, we will still face limitations due to the lack of resources to execute them one by one.

Reply:
> 3\. NDV

I don't understand. Could you please explain why it's important?